cklib/cklib/x509.py

```diff
@@ -0,0 +1,228 @@
+from cryptography.x509.base import Certificate, CertificateSigningRequest
+from typing import List, Optional, Tuple, Union
+
+
+def gen_rsa_key(key_size: int = 2048) -> RSAPrivateKey:
+    return rsa.generate_private_key(
+        public_exponent=65537, key_size=key_size, backend=default_backend()
+    )
+
+
+def bootstrap_ca(
+    days_valid: int = 3650,
+    common_name: str = "Cloudkeeper Certificate Authority",
+    organization_name: str = "Some Engineering Inc.",
+    locality_name: str = "San Francisco",
+    state_or_province_name: str = "California",
+    country_name: str = "US",
+) -> Tuple[RSAPrivateKey, Certificate]:
+    ca_key = gen_rsa_key()
+    subject = issuer = x509.Name(
+        [
+            x509.NameAttribute(NameOID.COUNTRY_NAME, country_name),
+            x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, state_or_province_name),
+            x509.NameAttribute(NameOID.LOCALITY_NAME, locality_name),
+            x509.NameAttribute(NameOID.ORGANIZATION_NAME, organization_name),
+            x509.NameAttribute(NameOID.COMMON_NAME, common_name),
+        ]
+    )
+    ca_cert = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(ca_key.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.utcnow().replace(tzinfo=timezone.utc))
+        .not_valid_after(
+            datetime.utcnow().replace(tzinfo=timezone.utc) + timedelta(days=days_valid)
+        )
+        .add_extension(x509.BasicConstraints(ca=True, path_length=None), critical=True)
+        .add_extension(
+            x509.KeyUsage(
+                digital_signature=False,
+                key_encipherment=False,
+                key_cert_sign=True,
+                key_agreement=False,
+                content_commitment=False,
+                data_encipherment=False,
+                crl_sign=True,
+                encipher_only=False,
+                decipher_only=False,
+            ),
+            critical=True,
+        )
+        .sign(ca_key, hashes.SHA256(), default_backend())
+    )
+    return ca_key, ca_cert
+
+
+def gen_csr(
+    csr_key: RSAPrivateKey,
+    common_name: str = "some.engineering",
+    san_dns_names: Optional[List[str]] = None,
+    san_ip_addresses: Optional[List[str]] = None,
+) -> CertificateSigningRequest:
+    if san_dns_names is None:
+        san_dns_names = []
+    if san_ip_addresses is None:
+        san_ip_addresses = []
+    csr_build = x509.CertificateSigningRequestBuilder().subject_name(
+        x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, common_name)])
+    )
+    if len(san_dns_names) + len(san_ip_addresses) > 0:
+        csr_build = csr_build.add_extension(
+            x509.SubjectAlternativeName(
+                [x509.DNSName(n) for n in san_dns_names]
+                + [x509.IPAddress(make_ip(i)) for i in san_ip_addresses]
+            ),
+            critical=False,
```

Comment:
why false?

Reply:
Because those are optional extensions and generally not critical if unsupported. As opposed to the base extensions above.